[1단계 - 출석 구현] 레오(김준영) 미션 제출합니다.

kjyyjk opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요? 
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
<!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
https://prolog.techcourse.co.kr/studylogs/3897

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
구현의 양과 시간 마감에 쫓겨 구현을 우선 시 하다 보니 
기본적인 코드 컨벤션을 지키지 못하고, 객체 지향적으로 작성하지 못했습니다.
저 스스로가 만족할 수 있을만큼 리팩터링을 할 시간이 없었어서 아쉬움이 많이 남습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

안녕하세요 제이~ 7기 레오입니다.
이번 출석 미션으로 만나뵙게 되어 반가워요.😃
잘부탁드립니다!

제이에게 여쭤보고싶은게 몇 가지 있어요.

- TDD 개발 방식
이번 미션에서는 페어인 젠슨과 함께 TDD로 개발해보기를 시도했는데. 그 과정이 너무 막막하고 힘들었던 경험이 되었어요.😭
처음에 '출석', '크루'와 같이 중요하고 눈에 띄는 도메인들을 TDD로 개발하는 것에는 큰 어려움이 없었던 거 같아요.
근데 머릿속으로만 그림을 그려보며 이런게 필요하지 않을까?하며 TDD를 적용하려다 보니 어느 순간 잘안된다고 느꼈던 거 같아요. 예를 들어 A 기능을 TDD로 구현하자고 하니 그 과정에서 B가 필요하다고 느껴졌고, 또 B를 구현하자고 하니 C가 필요하다고 느껴지고 이런 식이었던 거 같아요. 그래서 이번 미션 끝나고 나서 TDD로 개발하기 위해서는 어느 정도 책임 분리와 도메인 설계를 하고 들어가는게 좋겠다고 느꼈어요. 제이는 보통 TDD로 개발한다고 하면 어느 정도 설계를 하고 시작을 하시나요?

- Crew의 출석 기록 위치
미션 요구사항에 따라 크루의 출석 기록을 보관했어야했습니다. 현재 코드를 보면 Crew 객체가 AttendanceHistoryManger를 상태로 가지고, AttendanceHistoryManager가 해당 Crew의 모든 출석 기록을 보관중이게끔 설계를 하고 구현했습니다.
```
public class Crew {
    private final String name;
    private final AttendanceHistoryManager attendanceHistoryManager;

    public Crew(String name) {
        this.name =  name;
        this.attendanceHistoryManager = new AttendanceHistoryManager();
    }
...
}
```
```
public class AttendanceHistoryManager {
    private final Set<AttendanceHistory> attendanceHistories = new HashSet<>();
...
}

```

초반에는 괜찮다고 생각했으나 구현을 할 수록 Crew가 AttendanceHistoryManger에게 그대로 메서드 호출을 위임해야하는 경우가 많아졌고, 과연 Crew가 출석에 대한 로직을 가지고 있는 것이 적절한 설계일까하는 의구심이 들었어요.
```
Crew의 메서드 중 일부

public void addAttendanceResult(AttendanceHistory attendanceHistory) {
    attendanceHistoryManager.addAttendanceHistory(attendanceHistory);
}

public AttendanceHistory modifyAttendanceResult(AttendanceHistory attendanceHistory, LocalTime localTime) {
    return attendanceHistoryManager.modifyAttendanceResult(attendanceHistory, localTime);
}

public AttendanceHistory getAttendanceHistory(LocalDate localDate) {
    return attendanceHistoryManager.getAttendanceHistory(localDate);
}

public Map<AttendanceType, Integer> calculateAttendanceResult(LocalDate localDate) {
    return attendanceHistoryManager.calculateAttendanceResult(localDate);
}

public CrewStatus calculateCrewStatus(Map<AttendanceType, Integer> attendanceResult) {
    return attendanceHistoryManager.calculateCrewStatus(attendanceResult);
}
```
페어랑 잠깐 의논을 거치고나서는 각각의 Crew가 자신의 출석 기록을 상태로 가지는 것이 아닌, 크루별 출석 기록들을 상태로 가지는(Map<Crew, AttendanceHistories>와 같이) 도메인을 분리하는게 맞지 않을까?라는 결론을 내렸습니다.(만약 리팩터링한다면 이 별도의 도메인 객체 이름을 AttendanceHistoryManger라고 지었을 수도 있을 거 같아요.) 결국 리팩터링까지는 하지 못했으나 해당 건에 대하여 제이는 어떻게 생각하시는지 궁금해요!

- 출석 수정 로직
출석 수정 로직에 대해 페어와 다음과 같은 두가지 방법을 고민했었습니다.
1. 기존 출석 객체인 ```AttendanceHistory```의 attendanceTime 필드를 수정한다.
2. 기존 출석 객체를 삭제하고 새로운 출석 객체를 생성해 저장한다. 

저희는 2번으로 갈 경우 일을 두번하는 것이지 않을까?하는 의견도 있어 1번으로 구현했습니다.
하지만 1번으로 가니, 수정을 완료한 후 수정 이전의 출석 기록을 출력하려고 할 때 어쩔 수 없이 Dto를 도입해 복잡해지는 경우가 생겼습니다. 
2번으로 갔으면 일을 두번할 지라도 로직이 더 간결 명확해지고 불변성도 챙길 수 있었을 거 같아요.
제이는 이에 대해 어떻게 생각하시는 지 궁금해요!

kjyyjk committed:
docs: 초기 기능명세서 작성

kjyyjk committed:
docs: 출석 확인 세부 기능명세 작성

kjyyjk committed:
feat: 등교 시간을 토대로 출석, 지각, 결석 판별 구현

kjyyjk committed:
feat: 주어진 날짜를 바탕으로 공휴일 혹은 주말 판별

kjyyjk committed:
feat: 출석 결과를 저장

kjyyjk committed:
feat: 출석 수정 로직 구현

kjyyjk committed:
docs: 출석 확인 및 출석 수정 세부 기능명세 작성

kjyyjk committed:
rename: 출석 관련 클래스 및 변수명 수정

kjyyjk committed:
feat: 중복 출석 시 검증 구현

- Crew 객체의 책임을 명확히 하기 위해 AttendancHistory를 도입하였음

kjyyjk committed:
feat: 주말 또는 공휴일에 출석할 경우 예외 처리

kjyyjk committed:
feat: 출석 기록이 존재하지 않는다면 예외 발생

kjyyjk committed:
refactor: AttendanceType Enum 상수로 변경

kjyyjk committed:
feat: 기능 선택 구현

- 존재하지 않는 기능일 경우, 예외를 발생시킨다.

kjyyjk committed:
feat: 출석 확인 기능 구현

kjyyjk committed:
rename: ABSENT에서 ABSENCE로 상수 이름 변경

- ENUM 상수명을 명사형으로 통일하기 위해 변경하였다.

kjyyjk committed:
feat: 출석 수정 기능 구현

kjyyjk committed:
rename: 클래스명 수정

kjyyjk committed:
refactor: 출석 수정 로직 변경

kjyyjk committed:
feat: 크루 별 출석 기록 확인 기능 구현

kjyyjk committed:
feat: 출석, 지각, 결석 횟수 출력 기능 구현

kjyyjk committed:
feat: 제적 위험자 확인 후 출력

kjyyjk committed:
refactor: 시스템의 오늘 날짜를 2024년 12월로 변경

kjyyjk committed:
refactor: Controller의 메서드 분리

kjyyjk committed:
refactor: 기능 단위 별 메서드 분리

kjyyjk committed:
feat: 파일 읽은 후 크루 및 출석 기록 초기화

kjyyjk committed:
refactor: 유틸 클래스 private 생성자로 변경

kjyyjk committed:
docs: 크루 출석 기록 추가

kjyyjk committed:
feat: 캠퍼스 운영 시간 기준으로 출석 시간 검증 구현

kjyyjk committed:
feat: 출력 시 개행 추가

kang-hyungu commented on src/main/java/attendance/CurrentDate.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.time.LocalDate;
+
+public interface CurrentDate {
```
인터페이스로 추출한 이유가 있을까요?

kjyyjk replied:
LocalTime.now()는 제어하기가 힘든 부분이기도 하고, 기능 요구사항에 따라 2024년 12월을 기준으로 동작해야하기 때문에 인터페이스로 따로 분리했었던 거 같습니다.

덧붙이자면 '한정 운영 기간이 끝나고 나서는 CurrentDate를 구현하는 SystemCurrentDate만 다시 LocalTime.now()를 반환하도록 수정하면 변경에 유연하지 않을까?' 싶어 인터페이스로 분리했었습니다.

구현 당시에는 그러한 생각들을 가졌었지만, 지금 생각해보았을 때 어차피 한 곳에서 현재 날짜를 구해 주입해주는 구조이면 그냥 인터페이스 추출 없이 코드를 작성했어도 제가 생각했던 이유들이 충족되지않을까하는 생각도 드는 거 같아요.

kjyyjk replied:
덧붙이자면 '한정 운영 기간이 끝나고 나서는 CurrentDate를 구현하는 SystemCurrentDate만 다시 LocalTime.now()를 반환하도록 수정하면 변경에 유연하지 않을까?' 싶어 인터페이스로 분리했었습니다.

kang-hyungu commented on src/main/java/attendance/SystemCurrentDate.java:
```diff
@@ -0,0 +1,10 @@
+public class SystemCurrentDate implements CurrentDate {
+    @Override
+    public LocalDate now() {
+        return LocalDate.now().minusMonths(2);
```
클래스명만 보면 현재 시스템 날짜를 구하는 일을 맡긴 것 같은데, 왜 실제 로직에서는 2달을 빼는 건가요?

kjyyjk replied:
우선 12월 한정으로 시범 운영한다는 기능 요구사항이 있어 프로그램 상 현재 날짜는 2024년 12월이 되어야한다고 생각했습니다. 그렇기 때문에 현재 날짜인 2025년 2월에서 2달을 빼주어 2024년 12월로 반환하게끔 구현했습니다.

이렇게 했을 때 현재 날짜가 2025년 3월로 넘어 가버리면 결국 기능 요구사항과 달리 now()는 2025년 1월을 반환하게 된다는 문제를 가지고 있습니다. 따라서 추후에는 현재 날짜와 상관없이 고정적으로 2024년 12월 날짜를 반환하도록 리팩터링할 예정이었습니다.

kang-hyungu commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.List;
+import java.util.Map;
+
+public class AttendanceController {
```
컨트롤러라고 이름 지은 이유가 있을까요?

kjyyjk replied:
당시에는 단순히 MVC 패턴의 Controller이니 그렇게 지었던 거 같아요.

그리고 저는 Application의 main 메서드가 MVC 패턴 내 Controller의 역할을 충분히 수행할 수 있다고 생각하는 거 같아요. 즉 아직까지 별도의 Contoller를 만들어야 할 이유를 찾지 못하였기 때문에, Controller 내부의 로직을 Application으로 옮겼습니다.
b5d14cda7a013cda05af8d859f2c4bcf35c99942

kjyyjk replied:
혹시 더 적절한 이름이 있다면 추천해주실 수 있을까요?

kang-hyungu commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,101 @@
+
+    public void start() {
+        LocalDate now = currentDate.now();
+        while (true) {
```
if절이 너무 많네요.
새로운 조건이 생길 때마다 코드가 점점 길어질 것 같은데.. 더 나은 방법 없을까요?

kjyyjk replied:
옵션(String)을 key로, 옵션에 대한 동작(Runnable)을 value로 가지는 Map을 활용하도록 수정했습니다.
이렇게 하니 분기문도 줄어들어 코드를 읽기가 쉽고, 추후에 옵션이 추가 되더라도 유지보수하기 좋을 거 같아요!
e3ff044beb279147694f327a0897b4cd938a5603

kang-hyungu commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,101 @@
+
+    private Crew findCrew() {
+        String crewName = inputView.inputCrewName();
+        Crew crew = crewManager.findByCrewName(crewName);
```
```suggestion
        return crewManager.findByCrewName(crewName);
```

kang-hyungu commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,101 @@
+        LocalTime attendanceTime = inputView.inputAttendanceTime();
+        AttendanceType attendanceType = AttendancePolicy.checkAttendanceType(now, attendanceTime);
+        AttendanceHistory attendanceHistory = new AttendanceHistory(LocalDateTime.of(now, attendanceTime),
+                attendanceType);
```
왜 여기서 줄바꿈을 했을까요?

kjyyjk replied:
intellij에서 자동 포매팅을 했더니 코드가 너무 길어 줄바꿈이 들어간 거 같아요.
위에 피드백 주신 LocalDateTime.of를 변수로 추출했더니 길이가 줄어 줄바꿈이 필요없어졌습니다!
b09496d3d09bcc5de853a79a292fa4cdc5eeb176

kang-hyungu commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,101 @@
+        Crew crew = findCrew();
+        LocalTime attendanceTime = inputView.inputAttendanceTime();
+        AttendanceType attendanceType = AttendancePolicy.checkAttendanceType(now, attendanceTime);
+        AttendanceHistory attendanceHistory = new AttendanceHistory(LocalDateTime.of(now, attendanceTime),
```
LocalDateTime.of(now, attendanceTime)
이게 뭘 의미할까요? 적절한 변수명을 붙여주세요.

kjyyjk replied:
출석하려는 날짜와 시간 정보를 담고 있으니 attendanceDateTime이 적절하다고 생각되어 수정했습니다.
b09496d3d09bcc5de853a79a292fa4cdc5eeb176

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+    private final Set<AttendanceHistory> attendanceHistories = new HashSet<>();
+
+    public void addAttendanceHistory(AttendanceHistory attendanceHistory) {
+        if (!attendanceHistories.add(attendanceHistory)) {
```
한 줄에서 너무 많은 일을 하고 있네요.
한 줄에 한 의미만 표현하면 코드 읽기가 더 편해질거에요.

attendanceHistories.add(attendanceHistory)의 결과를 변수로 추출하고,
if절에서 사용해보시죠.

kjyyjk replied:
add의 결과는 Set에 추가되었는지 안되었는지, 즉 이미 출석이 존재하는지 여부를 의미하니 isAttendanceExists라는 변수로 추출하도록 수정했습니다.
훨씬 읽기 쉬워진 거 같아요!
29d54c37baa4c88fb6bae872060c7570b21fa0a1

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+
+    public AttendanceHistory getAttendanceHistory(LocalDate localDate) {
+        return attendanceHistories.stream()
+                .filter(history -> history.getAttendanceTime().toLocalDate().equals(localDate))
```
규칙 4: 한 줄에 점을 하나만 찍는다.

kjyyjk replied:
단순히 줄바꿈을 하기 보다는 ```AttendanceHistory```가 주어진 날짜와 자신의 출석 날짜가 같은지 여부를 반환하는 게 적절한 거 같아, ```AttendanceHistory```에 isAttendanceDateEquals 메서드를 생성하여 사용하도록 수정하였습니다.
29d54c37baa4c88fb6bae872060c7570b21fa0a1

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+
+    public AttendanceHistory modifyAttendanceResult(AttendanceHistory modifyAttendanceHistory, LocalTime localTime) {
+        AttendanceType attendanceType = AttendancePolicy.checkAttendanceType(
+                modifyAttendanceHistory.getAttendanceTime().toLocalDate(), localTime);
```
규칙 4: 한 줄에 점을 하나만 찍는다.

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+            LocalDate date = LocalDate.of(localDate.getYear(), localDate.getMonthValue(), i);
+            try {
+                AttendancePolicy.checkHoliday(date);
+            } catch (IllegalArgumentException e) {
```
이펙티브 자바 3판 참고하세요
아이템 69. 예외는 진짜 예외 상황에만 사용하라

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+            }
+            boolean flag = false;
+            for (AttendanceHistory attendanceHistory : attendanceHistories) {
+                if (attendanceHistory.getAttendanceTime().toLocalDate().equals(date)) {
```
규칙 4: 한 줄에 점을 하나만 찍는다.

kjyyjk replied:
위 피드백과 같이 isAttendanceDateEquals를 활용해 수정했습니다.
29d54c37baa4c88fb6bae872060c7570b21fa0a1

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+            }
+            boolean flag = false;
+            for (AttendanceHistory attendanceHistory : attendanceHistories) {
+                if (attendanceHistory.getAttendanceTime().toLocalDate().equals(date)) {
```
if절을 메서드로 추출해서 적절한 메서드명을 붙이면 좀더 코드 파악하기 용이할 것 같네요.

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+                if (attendanceHistory.getAttendanceTime().toLocalDate().equals(date)) {
+                    AttendanceType attendanceType = attendanceHistory.getAttendanceType();
+                    attendanceResult.put(attendanceType, attendanceResult.get(attendanceType) + 1);
+                    flag = true;
```
for문을 통해 출석 결과를 저장하면서 플래그를 처리하려다보니 코드가 복잡해진 것 같네요.
for문은 출석 결과만 저장하고, 다시 for문이나 stream을 사용해서 flag를 찾아내면 좀 더 코드를 파악하기 좋지 않을까요?

kjyyjk replied:
출석 결과 저장과 flag를 구하는 로직 모두 stream으로 구현해도 좋을 것 같아 둘다 stream을 사용하게끔 수정했습니다.
25637bb703631191aeb335e9e83f403ff1d86f02

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+    public CrewStatus calculateCrewStatus(Map<AttendanceType, Integer> attendanceResult) {
+        int validateValue = 0;
+        validateValue += attendanceResult.get(ABSENCE);
+        validateValue += attendanceResult.get(LATE) / 3;
```
3은 무엇을 의미할까요?

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+        return attendanceResult;
+    }
+
+    public CrewStatus calculateCrewStatus(Map<AttendanceType, Integer> attendanceResult) {
```
Map<AttendanceType, Integer>
요 구조가 제법 보이는데 일급 컬렉션으로 만들면 어떨까요?

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+        int validateValue = 0;
+        validateValue += attendanceResult.get(ABSENCE);
+        validateValue += attendanceResult.get(LATE) / 3;
+        if (validateValue > 5) {
```
5, 3, 2가 무엇을 의미하나요? 상수로 관리해주세요.

kang-hyungu commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,88 @@
+        return CLEAR;
+    }
+
+    private Map<AttendanceType, Integer> initializeAttendanceResult() {
```
요 메서드는 AttendanceType 클래스 내부에서 처리해도 괜찮을 것 같네요.

kang-hyungu commented on src/main/java/attendance/domain/AttendancePolicy.java:
```diff
@@ -0,0 +1,61 @@
+
+    public static void checkHoliday(LocalDate localDate) {
+        DayOfWeek dayOfWeek = localDate.getDayOfWeek();
+        if (dayOfWeek != DayOfWeek.SATURDAY && dayOfWeek != DayOfWeek.SUNDAY &&
```
```suggestion
        if (isNotSaturday(dayOfWeek) && isNotSunday(dayOfWeek) && isNotChristmas(localDate)) {
```
적절한 메서드로 감싸면 좀 더 코드 파악하기에 좋죠

kjyyjk replied:
해당 메서드에서는 주말 혹은 공휴일인지 체크하는 것이기 때문에 isNotWeekend와 isNotHoliday로 메서드를 추출했습니다.
1b2f39c306097711b128f16532cfa15803943814

kang-hyungu commented on src/main/java/attendance/domain/AttendancePolicy.java:
```diff
@@ -0,0 +1,61 @@
+    public static void checkHoliday(LocalDate localDate) {
+        DayOfWeek dayOfWeek = localDate.getDayOfWeek();
+        if (dayOfWeek != DayOfWeek.SATURDAY && dayOfWeek != DayOfWeek.SUNDAY &&
+                !localDate.equals(LocalDate.of(2024, 12, 25))) {
```
조건을 파악하기 쉽게 &&를 기준으로 줄바꿈을 하면 어떨까요?

```java
        if (dayOfWeek != DayOfWeek.SATURDAY &&
            dayOfWeek != DayOfWeek.SUNDAY && 
            !localDate.equals(LocalDate.of(2024, 12, 25))) {
```

kang-hyungu commented on src/main/java/attendance/domain/AttendancePolicy.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private static AttendanceType checkMondayAttendanceType(LocalTime localTime) {
+        if (localTime.isBefore(LocalTime.of(13, 6))) {
```
LocalTime.of(13, 6)이 무엇을 의미하나요?

kjyyjk replied:
월요일의 지각 기준 시간을 의미하는데,
static final 상수로 추출해 의미를 알기 쉬운 상수명을 지으면 좋을 거 같아요.
수정했습니다.
0b2f4a411d1fd36e1f200f48d6f11ba695f5b358

kjyyjk replied:
추가로 AttendancePolicy 내부에 캠퍼스 운영과 관련된 시간 상수들이 많아,
CampusRule이라는 Enum 클래스로 분리했습니다.
2f6d4cec23b3b876f317366d92824be7d700aaae

kang-hyungu commented on src/main/java/attendance/domain/Crew.java:
```diff
@@ -0,0 +1,57 @@
+    private final AttendanceHistoryManager attendanceHistoryManager;
+
+    public Crew(String name) {
+        this.name =  name;
```
크루 이름은 null 또는 빈값도 허용하나요?

kjyyjk replied:
시간이 부족해 놓쳤던 부분인 거 같아요. 수정했습니다.
aa6d97c3ffafc623215e46f5b908831116ca969d

kang-hyungu commented on src/main/java/attendance/domain/CrewManager.java:
```diff
@@ -0,0 +1,42 @@
+        for (Crew crew : crews) {
+            Map<AttendanceType, Integer> attendanceResult = crew.calculateAttendanceResult(localDate);
+            CrewStatus crewStatus = crew.calculateCrewStatus(attendanceResult);
+            if (crewStatus == CLEAR) {
```
2뎁스를 1뎁스로 줄일 방법 없을까요?

kjyyjk replied:
Stream을 사용하여 뎁스 문제를 해결하였습니다.
ddfcfcd31dc2b4a29d08208c4240f17375d4f75d

kang-hyungu commented on src/main/java/attendance/domain/dto/AttendanceHistoryDto.java:
```diff
@@ -0,0 +1,27 @@
+import attendance.domain.AttendanceType;
+import java.time.LocalDateTime;
+
+public class AttendanceHistoryDto {
```
dto가 필요한 상황인가요?

kjyyjk replied:
기존 코드는 출석 수정 시 필드 값을 변경해버리기 때문에 이전 값을 유지하기 위해 dto를 도입했었습니다.
하지만 필드 값을 변경하는 것이 아닌, 새로운 출석 객체를 만들어 저장하는 방식으로 리팩터링하였기 때문에
dto가 필요 없어 제거하였습니다. 
5779039a5600e6ce619aec7a702834fbc12f9fa4

kang-hyungu commented on src/main/java/attendance/utils/AttendanceFileReader.java:
```diff
@@ -0,0 +1,38 @@
+        return new BufferedReader(new FileReader(PATH));
+    }
+
+    public static void initializeAttendances(final BufferedReader br, CrewManager crewManager) throws IOException {
```
1. 메서드가 10줄을 넘어가지 않도록 작성해보세요.
2. br 줄여서 쓰지 말고 풀네임으로 작성해주세요.

kjyyjk replied:
수정했습니다!
32e715252cc003a17e633c9c0422fb0d9ee6cc5b

kang-hyungu commented on src/main/java/attendance/utils/AttendanceFileReader.java:
```diff
@@ -0,0 +1,38 @@
+        return new BufferedReader(new FileReader(PATH));
+    }
+
+    public static void initializeAttendances(final BufferedReader br, CrewManager crewManager) throws IOException {
```
예외를 Application으로 던지고 아무 처리도 안 하고 있네요.
예외에 대한 적절한 처리를 하시기 바랍니다.

이펙티브 자바 3판
아이템 77. 예외를 무시하지 말라

kang-hyungu commented on src/main/java/attendance/utils/AttendanceFileReader.java:
```diff
@@ -0,0 +1,38 @@
+            LocalDateTime dateTime = LocalDateTime.parse(split[1], formatter);
+            AttendanceHistory attendanceHistory = new AttendanceHistory(dateTime,
+                    AttendancePolicy.checkAttendanceType(dateTime.toLocalDate(), dateTime.toLocalTime()));
+            if (!crewManager.addCrew(crew)) {
```
2뎁스를 1뎁스로 줄일 방법 없나요?

kjyyjk replied:
크루가 add되었든 안되었든, 다시 crewName을 기반으로 크루를 조회하면 if문이 필요가 없겠네요!
수정했습니다.
32e715252cc003a17e633c9c0422fb0d9ee6cc5b

kang-hyungu commented on src/main/java/attendance/utils/AttendanceFileReader.java:
```diff
@@ -0,0 +1,38 @@
+            String nickname = split[0];
+            Crew crew = new Crew(nickname);
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+            LocalDateTime dateTime = LocalDateTime.parse(split[1], formatter);
```
split[1]에서 1이 뭔가요?

kjyyjk replied:
split 결과 내 출석 시간 정보의 인덱스이므로 INDEX_AS_ATTENDANCE_DATE_TIME라는 이름의 static 상수로 분리하였습니다.
32e715252cc003a17e633c9c0422fb0d9ee6cc5b

kang-hyungu commented on src/main/java/attendance/view/InputView.java:
```diff
@@ -0,0 +1,61 @@
+public class InputView {
+    private static final String TODAY_INFO = "오늘은 %d월 %d일 %s입니다. 기능을 선택해 주세요.";
+
+    private final Scanner sc = new Scanner(System.in);
```
약어 금지입니다.

kang-hyungu commented on src/main/java/attendance/view/InputView.java:
```diff
@@ -0,0 +1,61 @@
+        System.out.println("수정하려는 날짜(일)를 입력해 주세요.");
+        int endDate = localDate.lengthOfMonth();
+        String userInput = userInput();
+        if (!userInput.matches("\\d+")) {
```
숫자인지 아닌지 체크하는 로직을 둘 필요 없이, 그냥 int 타입으로 입력 받으면 되지 않나요?

kjyyjk replied:
scanner.nextInt()로 받으니 int인지 체크하는 로직과 int 변환하는 로직이 필요가 없네요!
수정했습니다.
b99f126b61695630958f11a7a398699cd6605231

kang-hyungu commented on src/main/java/attendance/view/InputView.java:
```diff
@@ -0,0 +1,61 @@
+        if (date < 1 || date > endDate) {
+            throw new IllegalArgumentException("잘못된 날짜입니다.");
+        }
+        return LocalDate.of(localDate.getYear(), localDate.getMonthValue(), date);
```
LocalDate.of(...)에서 어차피 잘못된 타입, 값을 입력하면 예외가 발생할테니 위에 검증 로직을 제거하고 try catch를 여기서 쓰면 코드량을 줄일 수 있겠네요.

kjyyjk replied:
LocalDate.of()에서 잘못된 날짜를 입력하면 DateTimeException이 발생하네요.
피드백주신대로 LocalDate.of를 try catch로 감싸주어 47-49라인의 검증 로직을 없앴습니다.
b99f126b61695630958f11a7a398699cd6605231

kang-hyungu commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,134 @@
+        );
+    }
+
+    public void printAttendanceHistories(LocalDate now, Crew crew) {
```
너무 길어서 코드 읽기 힘드네요
10줄 이하로 줄여보세요.

kang-hyungu commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,134 @@
+    public void printAttendanceResult(Map<AttendanceType, Integer> attendanceResult) {
+        for (AttendanceType attendanceType : attendanceResult.keySet()) {
+            System.out.println("%s: %d회".formatted(
+                    attendanceType,
```
여기서 왜 줄바꿈 한거에요?

kang-hyungu commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,134 @@
+        dangerousCrews.sort(new Comparator<Crew>() {
+            @Override
+            public int compare(Crew o1, Crew o2) {
+                CrewStatus crewStatus1 = o1.calculateCrewStatus(o1.calculateAttendanceResult(now));
```
변수명에 1, 2와 같은 의미 없는 이름을 붙이지 마세요.
비교하는 대상이 무엇인지 정확히 지칭하세요.

kang-hyungu reviewed (changes requested):
레오 몇 가지 피드백 남겼어요.
수정해보시고 리뷰 재요청주세요.
궁금한 점은 dm 주시거나 코치룸으로 찾아오세요~

kjyyjk committed:
refactor: 출석/지각/결석 기준 시간 상수화

- 기존에는 AttendancePolicy 내부에서 출석/지각/결석을 판별하기 위해 메서드 내에서
LocalTime.of를 사용하고 있었다. 하지만 단순히 코드를 보았을 때 이게 어떤 시간인지
알기가 쉽지가 않다. 따라서 static final 상수로 추출하여 적절한 네이밍을 해주고
가독성을 증진시킨다.

kjyyjk committed:
refactor: 캠퍼스 관련된 시간을 CampusRule로 Enum 상수화

- 기존 AttendancePolicy내에 있는 캠퍼스 관련된 시간 상수들을
CampusRule이라는 Enum 상수로 추출한다.

kjyyjk committed:
rename: 변수명으로 약어를 사용하지 않도록 수정

kjyyjk committed:
refactor: Controller에서 사용하는 변수명 및 메서드명 수정

- doAttendance()의 LocalTime.of는 출석하려는 날짜와 시간 정보를 담고 있으므로
attedanceDateTime 변수로 추출한다.

- controller에서 호출하고 있는 attendanceHistory 관련 메서드들의 명이
attendanceResult로 명명되어있는 것을 확인했다. 이는 전에 클래스명을 수정하면서
놓친 부분이기 때문에 마찬가지로 수정해준다.

kjyyjk committed:
refactor: 동일 날짜 체크 로직의 위치를 AttendanHistory로 변경

- 기존에는 외부에서 AttendanceHistory의 LocalDateTime을 받아와 이를
LocalDate로 변환하고 특정 날짜와 같은지를 확인했다. 하지만 이처럼 특정 날짜와
출석 날짜가 같은 지 확인하는 기능은 출석 시간을 상태로 가지고 있는
AttendanceHistory의 책임이라고 본다. 따라서 AttendanceHistory에
isAttendanceDateEquals()를 만들고 이를 테스트한다.

kjyyjk committed:
refactor: Application이 Controller의 역할을 하도록 변경

- 기존에는 별도의 Controller를 생성하여 사용했으나, 큰 이유가 존재하지 않았다.
그리고 Application의 main()도 충분히 MVC패턴의 Controller 역할을 잘 수행할 수
있다고 판단하였기 때문에 Controller를 삭제하고 모든 로직을 Applicatoin의
main() 내부로 옮긴다.

- 추후에 메서드를 잘게 분리할 예정이다.

kjyyjk committed:
refactor: 수정 날짜 입력 검증 로직 리팩터링

- 기존에는 문자열을 입력 받아 정수인지 확인하고, 정수로 변환하는 코드를 작성했다.
하지만 입력을 정수로 받는 하나의 로직으로 해결되기 때문에 수정한다.
만약 입력이 정수가 아니면 InputMismatchException이 발생하므로 try catch로
예외를 잡아 IllegalArgumentException으로 변환하여 던진다.

- 기존에는 입력 받은 날짜가 유효한 날짜인지 직접 확인했다. 하지만 LocalDate.of()로
결과를 반환하는 과정에서 잘못된 값이 들어가면 어차피 DateTimeException이 발생한다.
따라서 try catch로 예외를 잡아 IllegalArgumentException으로 변환하여 던진다.

kjyyjk committed:
refactor: OutputView 리팩터링

- 기존에는 한 메서드가 너무 길어 코드를 읽고 파악하기가 힘들다는 문제가 있었다.
따라서 적절하게 메서드를 나누어 코드의 가독성을 늘린다.

- 출력 문구 static final 상수화

kjyyjk committed:
refactor: AttendanceHistoryManager 리팩터링

- 기존 initiallizeAttendanceResult()를 AttendanceType의 책임으로 보고
AttendanceType 내부로 이동한다.

- 지각 3번이 결석 1번이랑 같다는 것을 의미하는 매직 넘버 3을 상수화한다.

- 기존에는 calculateAttendanceResult의 for문 내에 flag가 복잡하게 얽혀
있었다. 따라서 먼저 존재하는 출석에 대한 결과를 추가한 이후에.
따로 stream을 통해 출석이 존재하지 않는 지(flag)를 구한다.

- 하나의 메서드가 너무 많은 책임을 지지 않도록 메서드를 분리한다.

kjyyjk committed:
feat: 크루의 이름 검증 로직 추가 및 테스트 수행

- 크루의 이름이 null 또는 공백이면 예외를 발생하는 검증 로직을 추가한다.

kjyyjk committed:
refactor: checkHoliday() 리팩터링

- 주말 또는 공휴일이면 예외를 발생하는 메서드이기에,
checkNotWeekendAndHoliday로 메서드명을 변경한다.

- 내부 조건문들을 별도의 메서드인 isNotWeekend, isNotHoliday로 추출하여
가독성을 늘린다.

- 테스트 코드 또한 Parameterized Test를 수행한다.

kjyyjk committed:
refactor: AttendanceHistoryManager의 출석 로직 리팩터링

- 기존에는 출석을 하기 위해 미리 출석 객체인 AttendanceHistory를 생성하여
인자로 넘겼으나 어색하다고 느껴졌다. 따라서 출석할 날짜와 시간을 인자로 넘기면
Manager 내부에서 History를 생성하고, 검증 뒤 추가하게끔 수정한다.

kjyyjk committed:
refactor: AttendanceHistoryManager의 출석 수정 로직 리팩터링

- 기존에는 출석을 수정하기 위해 기존 출석 객체인 AttendanceHistory를 가져오고
인자로 출석 시간을 같이 넘겨주었다. 하지만 인자로 출석을 수정하려는 날짜와
출석을 수정하려는 시간만 건네주는 것이 더 자연스러운 협력이라고 생각되어
수정하였다.

kjyyjk committed:
refactor: 파일 읽어오는 로직 리팩터링

- 매직 넘버 및 패턴 상수화

- 메서드 분리

- 크루가 추가되었든 안되었든 다시 crewName을 기반으로 찾아와 출석을 초기화하면
while문 안에 if문이 필요가 없다.

kjyyjk committed:
refactor: 테스트 형식 통일 및 Parameterized Test 적용

kjyyjk committed:
refactor: 제적 위험자 확인 시 Stream 사용

kjyyjk committed:
rename: 변수명에 의미없는 숫자 제거

kjyyjk committed:
refactor: 각 옵션에 대한 동작을 Map으로 보관

- 기존에는 Application 내부에 각 옵션에 대한 분기문이 쭉
작성되어있었다. 이 경우에는 옵션이 추가될수록 코드의 길이가 점점 늘어나며
유지보수가 힘들어질 것이다. 때문에 옵션 문자를 키로 가지고,
옵션에 대한 행동인 Runnable을 값으로 가지는 Map을 사용한다.

kjyyjk committed:
fix: nextInt 후 잔여 개행문자 처리

- scanner.nextInt()로 수정 날짜를 입력받을 때 엔터로 인한
개행문자가 scanner에 남게된다. 따라서 다음 nextLine()시 입력으로
간주되어 프로그램이 오작동하는 것을 확인하였다. 이를 해결하기 위해
nextInt() 이후 잔여 개행을 해소하기 위한 nextLine()을 추가로 수행한다.

kjyyjk committed:
feat: IOException 예외 처리

- 기존에는 파일 입출력 시 발생하는 IOException을 그대로 Application에
던지고, Application에서도 아무런 조치 없이 예외를 던지고 있다.
잠재적인 문제 발생을 고려하여 우선 예외를 잡아 정해진 문구를 출력하도록
한다.

JunHoPark93 commented on src/main/java/attendance/domain/AttendanceHistoryManager.java:
```diff
@@ -0,0 +1,84 @@
+            try {
+                AttendancePolicy.checkNotWeekendAndHoliday(attendanceDate);
+            } catch (IllegalArgumentException e) {
+                continue;
```
예외를 catch해서 단순히 continue를 하면 어떤일이 벌어질까요

JunHoPark93 commented:
> 페어랑 잠깐 의논을 거치고나서는 각각의 Crew가 자신의 출석 기록을 상태로 가지는 것이 아닌, 크루별 출석 기록들을 상태로 가지는(Map<Crew, AttendanceHistories>와 같이) 도메인을 분리하는게 맞지 않을까?라는 결론을 내렸습니다.(만약 리팩터링한다면 이 별도의 도메인 객체 이름을 AttendanceHistoryManger라고 지었을 수도 있을 거 같아요.) 결국 리팩터링까지는 하지 못했으나 해당 건에 대하여 제이는 어떻게 생각하시는지 궁금해요!

실생활에서도 출석을 관리하는 사람이 따로 있는경우를 종종 보셨을거에요. 이와 마찬가지로 역할을 분리를하는게 crew본인의 역할을 다하는데 더 좋지 않을까하는 생각입니다. 현재는 manager를 따로 두셨네요! 잘하신거같습니다.

> 출석 수정 로직
> 출석 수정 로직에 대해 페어와 다음과 같은 두가지 방법을 고민했었습니다.
> 기존 출석 객체인 AttendanceHistory의 attendanceTime 필드를 수정한다.
> 기존 출석 객체를 삭제하고 새로운 출석 객체를 생성해 저장한다.
> 저희는 2번으로 갈 경우 일을 두번하는 것이지 않을까?하는 의견도 있어 1번으로 구현했습니다.
> 하지만 1번으로 가니, 수정을 완료한 후 수정 이전의 출석 기록을 출력하려고 할 때 어쩔 수 없이 Dto를 도입해 복잡해지는 경우가 생겼습니다.
> 2번으로 갔으면 일을 두번할 지라도 로직이 더 간결 명확해지고 불변성도 챙길 수 있었을 거 같아요.
> 제이는 이에 대해 어떻게 생각하시는 지 궁금해요!

1번으로 했을때 이전상태를 가지고있으면 되는것아닌가요? 제가 질문을 잘 이해를 못한걸수도있는데 이부분 해결안된거라면 다음 단계에서 다시 언급주세요

JunHoPark93 commented:
> TDD 개발 방식
> 이번 미션에서는 페어인 젠슨과 함께 TDD로 개발해보기를 시도했는데. 그 과정이 너무 막막하고 힘들었던 경험이 되었어요.😭
> 처음에 '출석', '크루'와 같이 중요하고 눈에 띄는 도메인들을 TDD로 개발하는 것에는 큰 어려움이 없었던 거 같아요.
> 근데 머릿속으로만 그림을 그려보며 이런게 필요하지 않을까?하며 TDD를 적용하려다 보니 어느 순간 잘안된다고 느꼈던 거 같아요. 예를 들어 A 기능을 TDD로 구현하자고 하니 그 과정에서 B가 필요하다고 느껴졌고, 또 B를 구현하자고 하니 C가 필요하다고 느껴지고 이런 식이었던 거 같아요. 그래서 이번 미션 끝나고 나서 TDD로 개발하기 위해서는 어느 정도 책임 분리와 도메인 설계를 하고 들어가는게 좋겠다고 느꼈어요. 제이는 보통 TDD로 개발한다고 하면 어느 정도 설계를 하고 시작을 하시나요?

보통 설계할때는 발생할 수 있는 경우의 수들을 코딩하기전에 미리 정리하고 가는데요, 코딩을 하다가도 말씀주신 a,b,c가 계속 떠오른다면 설계를 부족하게 한것이니 코딩 (tdd든 아니든)을 멈추고 다시 문서를 정리합니다. 

저는 TDD를 많이 하지는 않아서 답을 명확히 드리기는 어렵지만, (혼자해도 쉽지않은데 둘이 하게 되면 이야기가 가끔 산으로 가기도 했던거같아요.) 만약 TDD로 한다고했어도 위와같이 진행했을것 같습니다. 

혹시 답변이 부족했다면 다시 알려주세요~

JunHoPark93 commented:
구구가 이미 많은 리뷰를 해주셔서 저는 질문주신 부분들에 대해서만 코멘트를 남겼어요~ 혹시 다른 질문이 생기면 언제든 다시 말씀주세요

JunHoPark93 reviewed (approved):
안녕하세요 레오~
많은 리뷰 반영하시느라 고생많으셨네요
다음 단계에서 뵙겠습니다
